java/java.hints/src/org/netbeans/modules/java/hints/infrastructure/JavaErrorProvider.java

```diff
@@ -200,25 +202,55 @@ private static List<CodeAction> convertFixes(ErrorDescription err, Consumer<Exce
             }
         }
 
-        List<Fix> fixes = sortFixes(lfl.getFixes());
+        List<Fix> baseFixes = lfl.getFixes();
+
+        baseFixes = sortFixes(baseFixes);
+
+        List<Pair<Fix, Boolean>> fixesAndSubfixes = new ArrayList<>(); // true if main fix
+
+        for (Fix f : baseFixes){
+            fixesAndSubfixes.add(Pair.of(f, true));
+
+            try {
+                Class<?> controller = Class.forName("org.netbeans.modules.editor.hints.HintsControllerImpl",
+                                                    false,
+                                                    ErrorDescription.class.getClassLoader());
+                Method getSubfixes = controller.getDeclaredMethod("getSubfixes", Fix.class);
```

Comment:
calling `Class.forName` in a loop is not ideal. The method field could be set in a static initializer, and the loop would run when it isn't null.

I haven't looked at this in more detail but the fact that this needs reflection is a bit of a red flag already.

Reply:
If reflection is something that we don't want to use here, then other option is importing`spi.editor.hints` and then using method `HintsControllerImpl`. If that is fine then code can be updated accordingly.

Reply:
I am afraid there are only bad solutions:
- either the reflection (where the problem is obvious),
- implementation dependency between `java.hints` and `spi.editor.hints`, which is problematic as a) it goes across cluster boundary; b) `java.hints` is a purely hint implementation module, it should not "normally" need to hack into any implementation,
- putting `getSubfixes` to the API, which is problematic, as `spi.editor.hints` does not have any API currently, and there's so far only a single use of this, and that is here,
- maybe some tricks with moving `JavaErrorProvider` into a bridge module which would have impl dependency on both `java.hints` and `spi.editor.hints`, which also seem pretty ugly.

So, the reflection seemed least bad out of the options. It does not open any new avenues to get additional access, and it is a clear hack that will be obvious (as opposed to e.g. implementation dependency). But I admit my opinion here may not be the only one, or even the best.

Reply:
if it is decided that the reflection bits stay I would recommend the following:
 - move all relfective setup code into static initializer, log exception + descriptive msg, e.g "implementation changed, update needed"
 - also add a short comment that that this is the most convenient way to access the method without having to change module dependencies or complicate things further
 - inside the loop I would call a utility method, which simply returns `getSubfixes.invoke(null, f))` and deals with the reflective invocation exception there (maybe log an return empty iterable)
 
this would move the `Class.forName` out of inner loops and help future maintainers to fix it it again once things start failing

Reply:
> putting getSubfixes to the API, which is problematic, as spi.editor.hints does not have any API currently,

To clarify, because it doesn't have any API packages as opposed to public packages?

Reply:
`spi.editor.hints` only has SPI, which means something providers (that produce information) plug into, but that is not used to retrieve information. We could add an API package, with an API class, it just feels weird that the only method would be to get sub-fixes.

But maybe we could add a default method to `org.netbeans.spi.editor.hints.Fix`, e.g. `getSubfixes()`. `org.netbeans.modules.editor.hints.borrowed.ListCompletionView` would then use `Fix.getSubfixes()` to get sub-fixes instead of the current code, so that `Fix` subclasses could provided sub-fixes using the new method as well. And `java.hints` could also call that method. That's not completely clear, but would seem acceptable.

Reply:
I tried out this approach what @lahodaj suggested it seems to work fine, important point is it doesn't use reflection code. Updated the code accrodingly.